### ISSUE TITLE
Consul-cluster module instances don't elect a leader

### ISSUE THREAD

**Issue by michael-henderson** [hashicorp/terraform-aws-consul#21]
I have used packer to build multiple versions of consul/dnsmasq into an AMI. With every version the instances spin up, but do not elect a leader.  I have also tried the test with the example AMI that was provided on this page: https://registry.terraform.io/modules/hashicorp/consul/aws/0.0.5 Same result, no leader.  I've confirmed that the instances can reach each other on all ports defined in the module, so it's not a rogue security group rule or anything like that.

I'm not sure if I'm missing something, or the module just doesn't work as expected, but I'm open to any suggestions.

**Comment by brikis98** [comment]
What exactly are you deploying? For example, I just tried the following:

```bash
# Check out the code
git clone git@github.com:hashicorp/terraform-aws-consul.git
cd terraform-aws-consul

# Deploy
terraform init
terraform plan
terraform apply

# See what's running
./examples/consul-examples-helper/consul-examples-helper.sh 

Your Consul servers are running at the following IP addresses:

    52.201.182.177
    52.90.250.243
    34.236.190.70

Some commands for you to try:

    consul members -http-addr=52.201.182.177:8500
    consul kv put -http-addr=52.201.182.177:8500 foo bar
    consul kv get -http-addr=52.201.182.177:8500 foo

To see the Consul UI, open the following URL in your web browser:

    http://52.201.182.177:8500/ui/

# Try one of the commands
consul members -http-addr=52.201.182.177:8500

Node                 Address             Status  Type    Build  Protocol  DC         Segment
i-0520fb1ad9ad4da86  172.31.88.40:8301   alive   server  0.9.3  2         us-east-1  <all>
i-094d32c93ae418c24  172.31.6.145:8301   alive   server  0.9.3  2         us-east-1  <all>
i-0fb297a4dfcbbdd93  172.31.32.212:8301  alive   server  0.9.3  2         us-east-1  <all>
i-03230d166c1aef31c  172.31.66.230:8301  alive   client  0.9.3  2         us-east-1  <default>
i-04b9ff314ab137a41  172.31.61.203:8301  alive   client  0.9.3  2         us-east-1  <default>
i-0ba0b2cc627e0f270  172.31.83.61:8301   alive   client  0.9.3  2         us-east-1  <default>
i-0d72bbdc9593a8050  172.31.1.236:8301   alive   client  0.9.3  2         us-east-1  <default>
i-0eedcfe78b32b70ac  172.31.40.200:8301  alive   client  0.9.3  2         us-east-1  <default>
i-0f9862e0b60039914  172.31.25.122:8301  alive   client  0.9.3  2         us-east-1  <default>
...

**Comment by michael-henderson** [comment]
Thank you for the response!  I ran exactly what you did above, but when I got to running the helper script it just infinitely loops with the following output:

`$ ./examples/consul-examples-helper/consul-examples-helper.sh
2017-10-31 13:38:10 [INFO] [consul-examples-helper.sh] Looking up public IP addresses for 3 Consul server EC2 Instances.
2017-10-31 13:38:11 [INFO] [consul-examples-helper.sh] Fetching public IP addresses for EC2 Instances in us-west-2 with tag consul-servers=consul-example
2017-10-31 13:38:12 [INFO] [consul-examples-helper.sh] Found all 3 public IP addresses!
2017-10-31 13:38:12 [INFO] [consul-examples-helper.sh] Waiting for 3 Consul servers to register in the cluster
2017-10-31 13:38:12 [INFO] [consul-examples-helper.sh] Running 'consul members' command against server at IP address null
Error retrieving members: Get http://null:8500/v1/agent/members?segment=_all: dial tcp 92.242.140.2:8500: i/o timeout
2017-10-31 13:38:42 [INFO] [consul-examples-helper.sh] 1 out of 3 Consul servers have registered in the cluster.
2017-10-31 13:38:42 [INFO] [consul-examples-helper.sh] Sleeping for 10 seconds and will check again.
2017-10-31 13:38:52 [INFO] [consul-examples-helper.sh] Running 'consul members' command against server at IP address null`

**Comment by brikis98** [comment]
How odd:

```
Found all 3 public IP addresses!
```

But then:

```
Running 'consul members' command against server at IP address null
```

Why would it be `null`? 

Is it possible that you have disabled public IPs in your default VPC?

**Comment by michael-henderson** [comment]
Aha!  I was deploying into an existing VPC, and some subnets have public IP allocation on and others don't (based on networking layout).  The way instances were created, only one had a public IP attached to it.  Is it possible to use this module in a VPC without public IP addresses?  I'd prefer not to have them on my consul cluster anyway, TBH.

Side note: Might want to change the `Found all 3 public IP addresses!` output at some point, as it clearly didn't find all 3 :P

**Comment by michael-henderson** [comment]
Interesting... I tore everything down, and rebuilt and now the helper script finishes and comes back with a little more info:

```
$ ./examples/consul-examples-helper/consul-examples-helper.sh
2017-10-31 14:53:22 [INFO] [consul-examples-helper.sh] Looking up public IP addresses for 3 Consul server EC2 Instances.
2017-10-31 14:53:22 [INFO] [consul-examples-helper.sh] Fetching public IP addresses for EC2 Instances in us-west-2 with tag consul-servers=consul-example
2017-10-31 14:53:23 [INFO] [consul-examples-helper.sh] Found all 3 public IP addresses!
2017-10-31 14:53:24 [INFO] [consul-examples-helper.sh] Waiting for 3 Consul servers to register in the cluster
2017-10-31 14:53:24 [INFO] [consul-examples-helper.sh] Running 'consul members' command against server at IP address <public IP>
2017-10-31 14:53:24 [INFO] [consul-examples-helper.sh] All 3 Consul servers have registered in the cluster!

Your Consul servers are running at the following IP addresses:

    <public IP>
    null
    null

Some commands for you to try:

    consul members -http-addr=<public IP>:8500
    consul kv put -http-addr=<public IP>:8500 foo bar
    consul kv get -http-addr=<public IP>:8500 foo

To see the Consul UI, open the following URL in your web browser:

    http://<public IP>:8500/ui/
```

**Comment by brikis98** [comment]
Ah, OK, so the helper script definitely has a bug where it's not detecting the `null` values properly. We should fix that!

Now, if you SSH to the servers (or at least the one with a public IP) and run `consul members`, what do you see? Are they able to elect a leader or not? If not, the most likely reason is that your subnets are not able to talk to each other. Are the route tables configured correctly? Are there Network ACLs blocking traffic?

**Comment by brikis98** [comment]
BTW, by default, the cluster nodes talk to each other over private IPs, not public IPs: https://github.com/hashicorp/terraform-aws-consul/blob/master/modules/run-consul/run-consul#L92. The public IP portion is only so you can conveniently test the code from home.

**Comment by michael-henderson** [comment]
Looks like they are able to elect a leader :)

**Comment by brikis98** [comment]
Yup! In production deployments, we usually deploy solely into private subnets and don't use public IPs at all. They are just used in the "root example" in this module so it's easier to get started... Well, as long as your default VPC has the default settings that is :)

I'm going to rename this issue to "consul-examples-helper.sh doesn't check for public IPs properly" and mark it as a bug, as that script should be fixed to show a better error message.

**Comment by michael-henderson** [comment]
Thanks for the help, @brikis98  :)
